### PR TITLE
First proposal for US Sensors

### PR DESCRIPTION
**Pull request by carsten-kuebler**
Basic definition for use of US sensors. New detections and additional information for detected objects

### CODE REVIEW

**Comment by carsten-kuebler** [comment]
@CarloVanDriestenBMW Figures will be updated to .svg figures.

File: osi_featuredata.proto [hunk 1]
```diff
@@ -144,6 +160,74 @@ message SensorDetectionHeader
...
+        EXTENDED_QUALIFIER_INTERNAL_REASON = 10;
+
+        // External disturbance, sensor specific for front radar (e.g.
+        // Interference of different radar sensors)
+        //
+        EXTENDED_QUALIFIER_EXTERNAL_DISTURBANCE = 11;
+
+        // Beginning blockage, sensor specific for front radar
+        //
+        EXTENDED_QUALIFIER_BEBINNING_BLOCKAGE = 12;
```
Comment (CarloVanDriestenBMW) [inline osi_featuredata.proto]: typo?

File: osi_featuredata.proto [hunk 2]
```diff
@@ -104,13 +113,20 @@ message SensorDetectionHeader
...
-    // The ID of the sensor at host vehicle's mounting_position.
+    // The ID of the sensor at host vehicle's \c #mounting_position.
     //
     // This ID can equal \c SensorData::sensor_id, if \c SensorData holds only
     // data from one sensor/sensor model.
     //
     optional Identifier sensor_id = 7;
 
+    // The extended qualifier describes the reason (not the effect) why the
+    // event data qualifier \c #data_qualifier is reduced or not available.
```
Comment (CarloVanDriestenBMW) [inline osi_featuredata.proto]: [...] event data, indicated by the qualifier \c # data_qualifier, is reduced or not available.

File: osi_featuredata.proto [hunk 3]
```diff
@@ -104,13 +113,20 @@ message SensorDetectionHeader
...
+    //
+    // \note required for USS
```
Comment (CarloVanDriestenBMW) [inline osi_featuredata.proto]: Capital letter at the beginning and dot at the end? Please check all comments.

File: osi_featuredata.proto [hunk 4]
```diff
@@ -144,6 +160,74 @@ message SensorDetectionHeader
...
+        // Sensor invalid.
+        //
+        // \note required for USS
+        //
+        DATA_QUALIFIER_INVALID = 7;
+    }
+
+    //
+    // Extended qualifier describes the reason (not the effect) why the event
+    // data qualifier \c #data_qualifier is reduced or not available.
```
Comment (CarloVanDriestenBMW) [inline osi_featuredata.proto]: see above

File: osi_featuredata.proto [hunk 5]
```diff
@@ -144,6 +160,74 @@ message SensorDetectionHeader
...
+    {
+        // Unknown (must not be used in ground truth).
+        //
+        EXTENDED_QUALIFIER_UNKNOWN = 0;
+
+        // Other (unspecified but known).
+        //
+        EXTENDED_QUALIFIER_OTHER = 1;
+
+        // Normal operation mode
```
Comment (CarloVanDriestenBMW) [inline osi_featuredata.proto]: dots missing? check all comments

File: osi_featuredata.proto [hunk 6]
```diff
@@ -316,6 +400,154 @@ message LidarDetection
...
+//
+// Sending: Ultrasonic Sensor ID 1
+// Receiving: Ultrasonic Sensor ID 1
+// \image html OSI_USSensor_direct.png
+//
+// Indirect detections:
+//
+// Sending: Ultrasonic Sensor ID 1
+// Receiving: Ultrasonic Sensor ID 2 or/and 3
+// \image html OSI_USSensor_indirect.png
```
Comment (CarloVanDriestenBMW) [inline osi_featuredata.proto]: TODO: .svg as mentioned in PR

File: osi_featuredata.proto [hunk 7]
```diff
@@ -316,6 +400,154 @@ message LidarDetection
...
+    // Additional header attributes of ultrasonic detection from one ultrasonic
+    // sensor.
+    //
+    optional UltrasonicDetectionSpecificHeader specific_header = 3;
+
+    // List of ultrasonic detections.
+    //
+    // \note OSI uses singular instead of plural for repeated field names.
+    //
+    repeated UltrasonicDetection detection = 2;
```
Comment (CarloVanDriestenBMW) [inline osi_featuredata.proto]: Enumeration order on purpose?

File: osi_featuredata.proto [hunk 8]
```diff
@@ -316,6 +400,154 @@ message LidarDetection
...
+//
+message UltrasonicDetection
+{
+    // Existence probability of the detection not based on history. Value does
+    // not depend on any past experience with similar detections.
+    //
+    // \note Used as confidence measure where a low value means less confidence
+    // and a high value indicates strong confidence.
+    //
+    optional double existence_probability = 1;
```
Comment (CarloVanDriestenBMW) [inline osi_featuredata.proto]: please add value range here!

File: osi_featuredata.proto [hunk 9]
```diff
@@ -316,6 +400,154 @@ message LidarDetection
...
+
+    // Measured distance (radius) of the detection.
+    //
+    // Unit: [m]
+    //
+    optional double distance = 3;
+}
+
+//
+// \brief Ultrasonic detection received by an other ultrasonic sensor (different
```
Comment (CarloVanDriestenBMW) [inline osi_featuredata.proto]: another?

File: osi_featuredata.proto [hunk 10]
```diff
@@ -316,6 +400,154 @@ message LidarDetection
...
+//
+message UltrasonicIndirectDetection
+{
+    // Existence probability of the detection not based on history. Value does
+    // not depend on any past experience with similar detections.
+    //
+    // \note Used as confidence measure where a low value means less confidence
+    // and a high value indicates strong confidence.
+    //
+    optional double existence_probability = 1;
```
Comment (CarloVanDriestenBMW) [inline osi_featuredata.proto]: value range

File: osi_featuredata.proto [hunk 11]
```diff
@@ -316,6 +400,154 @@ message LidarDetection
...
+
+    // ID of the detected object this detection is associated to.
+    //
+    // \note ID = MAX(uint64) indicates no reference to an object.
```
Comment (CarloVanDriestenBMW) [inline osi_featuredata.proto]: invalid IDs should be treated the same way in all OSI ID messages? Please add issue to check all other IDs

File: osi_featuredata.proto [hunk 12]
```diff
@@ -316,6 +400,154 @@ message LidarDetection
...
+    //
+    optional Identifier object_id = 2;
+
+    // First parameter b of an ellipsoid equation.
+    //
```
Comment (CarloVanDriestenBMW) [inline osi_featuredata.proto]: Please add equation definition here

Comment (CarloVanDriestenBMW) [reply osi_featuredata.proto]: or reference

File: osi_featuredata.proto [hunk 13]
```diff
@@ -316,6 +400,154 @@ message LidarDetection
...
+    // Unit: [m]
+    //
+    optional double ellipsoid_axial = 4;
+
+    // The ID of the sensor's receiver. Sender ID is stored in the header \c
+    // SensorDetectionHeader.
+    //
+    optional Identifier receiver_id = 5;
+
+    // The vector to the receiver's origin in sending ultrasonic sensor frame.
```
Comment (CarloVanDriestenBMW) [inline osi_featuredata.proto]: Add a .svg image as explanation here?

File: osi_sensorspecific.proto [hunk 14]
```diff
@@ -43,6 +45,102 @@ message CameraSpecificObjectData
...
+    // of this object.
+    //
+    // Unit: [m]
+    //
+    optional double maximum_measurement_distance_sensor = 1;
+
+    // This value indicates the propability height for the classification in the
+    // calculated object.
+    //
+    // Unit: [%]
```
Comment (CarloVanDriestenBMW) [inline osi_sensorspecific.proto]: Normally it is a value [0,1] for all other probailities!

File: osi_sensorspecific.proto [hunk 15]
```diff
@@ -43,6 +45,102 @@ message CameraSpecificObjectData
...
+    //
+    optional double probability = 2;
+
+    // This indicates if the feature was calculated based on one or multiple
+    // sensors.
+    //
+    optional TrilaterationStatus trilateration_status = 3;
+
+    // Describes the general moving direction of the feature.
```
Comment (CarloVanDriestenBMW) [inline osi_sensorspecific.proto]: of the feature or detection?

File: osi_sensorspecific.proto [hunk 16]
```diff
@@ -43,6 +45,102 @@ message CameraSpecificObjectData
...
+    //
+    optional Trend trend = 4;
+
+    // Ultrasonic signalway. Snder to receiver.
```
Comment (CarloVanDriestenBMW) [inline osi_sensorspecific.proto]: typo

File: osi_sensorspecific.proto [hunk 17]
```diff
@@ -43,6 +45,102 @@ message CameraSpecificObjectData
...
+    //
+    // \note This information can also be derived from the corresponding \c
+    // UltrasonicDetection. \c UltrasonicDetection refer to DetectedXXX by \c
+    // UltrasonicDetection::object_id.
+    //
+    repeated Signalway singalway = 5;
```
Comment (CarloVanDriestenBMW) [inline osi_sensorspecific.proto]: typo

File: osi_sensorspecific.proto [hunk 18]
```diff
@@ -43,6 +45,102 @@ message CameraSpecificObjectData
...
+    {
+        // Unknown (must not be used in ground truth).
+        //
+        TRILATERATION_STATUS_UNKNOWN = 0;
+
+        // Other (unspecified but known).
+        //
+        TRILATERATION_STATUS_OTHER = 1;
+
+        // No trilateration used
```
Comment (CarloVanDriestenBMW) [inline osi_sensorspecific.proto]: dots missing

File: osi_sensorspecific.proto [hunk 19]
```diff
@@ -43,6 +45,102 @@ message CameraSpecificObjectData
...
+
+        // Other (unspecified but known).
+        //
+        TREND_OTHER = 1;
+
+        // \note There is no definition
+        //
+        TREND_CONSTANT_APPROACHING = 2;
+
+        // \note There is no definition
```
Comment (CarloVanDriestenBMW) [inline osi_sensorspecific.proto]: really? Please get in contact with the one who requested this property to get a definition.

File: osi_featuredata.proto (hunk 3 above)
Comment (carsten-kuebler) [reply osi_featuredata.proto]: I removed the note.

**Comment by carsten-kuebler** [comment]
@CarloVanDriestenBMW Update Trend definiton

File: osi_featuredata.proto (hunk 7 above)
Comment (CarloVanDriestenBMW) [reply osi_featuredata.proto]: @carsten-kuebler ping

**Pull request merged**